joomla/joomla-cms
# [4.0] Move Aurora to CSS Grid

### Summary of Changes
CSS grid decouples HTML from layout, allowing layout to be moved exclusively to CSS. This PR moves the Aurora frontend template layout from Bootstrap to CSS grid (https://www.w3.org/TR/css3-grid-layout/). 

### Testing Instructions
Apply patch and ensure frontend template displays correctly across all browsers. 

### Compatibility
Includes a flexbox fallback for browsers that do not support the new CSS grid spec, currently this means IE11 and Edge which only supports the older spec. Microsoft are currently working on adding the new spec to Edge. Considering the collapsing nature of Joomla layouts the older CSS grid spec is not really suitable to use as it relies on a predefined number of rows.

### Considerations
**Move layout controls to template style** - CSS grid allows us to easily set the layout in a single location with a few lines of CSS with the grid-template-areas property (https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout/Grid_Template_Areas). This opens up the possibility of creating a ‘layout builder’ of sorts which would be set from within the template style settings. A basic implementation would be just a textarea field however ideally we could improve on this with some kind of GUI. The textarea field would contain something like the following which can then be edited by the user…

```
head head head head
banner banner banner banner
top-a top-a top-a top-a
top-b top-b top-b top-b
side-l main-t main-t side-r
side-l comp comp side-r
side-l main-b main-b side-r
bot-a bot-a bot-a bot-a
bot-b bot-b bot-b bot-b
footer footer footer footer 
```

**Module Groups** - Considering the layout flexibility that CSS grid adds, we may need to reconsider the definition of a ‘Module Position’. ‘Module Groups’ may be more accurate.
 
**IE11, Expect a difference** - IE11 will continue to display as it does currently however as CSS grid features are implemented we need to expect some layout differences between IE11 and grid supporting browsers. As long as IE11 remains usable I can’t see this as a problem. An example of this can been seen with this PR if you resize the browser. Excluding IE11, on smaller screens the component area is moved above the top-# module positions. Polyfills do exist for IE11 however I am not sure if we want to go down that road. 
 
**Component views** - With a bit of imagination CSS grid opens up layout possibilities for the default core component views. It also partly removes the Bootstrap dependency. Whatever the thoughts on using CSS grid in the views, I believe that they should at least be CSS grid ready. This means removing Bootstrap grid classes. Bootstrap relies on nested divs with a tightly coupled class structure to define a grid which does not mix at all with CSS grid. 
 
### Misc.
**Nested Elements** - CSS grid uses a flat markup (no nesting). As a flexbox fallback is needed for IE11, I had to nest some of the elements (component, sidebar and main positions). Although this does restrict the freedom offered by CSS grid, it’s effect is minimal considering the current Aurora layout.
 
**No media queries** - By way of demonstration I used the minmax function instead of media queries for the top and bottom module widths.
 
**Auto module widths** - Top and bottom position module widths are calculated automatically depending on the number of modules in a position so as to allow modules to remain in a row with equal widths. Modules will wrap to a second row if widths go below a set px width (180px).

~~**Linting** -  Currently SCSS-Lint does not fully support CSS grid so was disabled to compile this PR.~~
